tests/pass/libc.rs

```diff
@@ -1,17 +1,142 @@
+    // Create file.
+    drop(File::create(&path).unwrap());
+    unsafe {
+        let r = libc::realpath(c_path.as_ptr(), std::ptr::null_mut());
+        assert!(!r.is_null());
+        buf = CStr::from_ptr(r).to_bytes().to_vec();
+        libc::free(r as *mut _);
+    }
+    let canonical = PathBuf::from(OsString::from_vec(buf));
+    // Cleanup after test.
+    remove_file(&path).unwrap();
+
+    assert_eq!(path.file_name(), canonical.file_name());
+}
+
+/// Test non-allocating variant of `realpath`.
+fn test_posix_realpath_noalloc() {
+    use std::ffi::{CStr, CString};
+    use std::fs::{remove_file, File};
+    use std::os::unix::ffi::OsStrExt;
+    use std::path::PathBuf;
+
+    let path = tmp().join("miri_test_libc_posix_realpath_noalloc");
+    let c_path = CString::new(path.as_os_str().as_bytes()).expect("CString::new failed");
+
+    let mut v = vec![0; libc::PATH_MAX as usize];
+
+    // Cleanup before test.
+    remove_file(&path).ok();
+    // Create file.
+    drop(File::create(&path).unwrap());
+    unsafe {
+        let r = libc::realpath(c_path.as_ptr(), v.as_mut_ptr());
+        assert!(!r.is_null());
+    }
+    let c = unsafe { CStr::from_ptr(v.as_ptr()) };
+    let canonical = PathBuf::from(c.to_str().expect("CStr to str"));
+    // Cleanup after test.
+    remove_file(&path).unwrap();
+
+    assert_eq!(path.file_name(), canonical.file_name());
+}
+
+/// Test failure cases for `realpath`.
+fn test_posix_realpath_errors() {
+    use std::convert::TryInto;
+    use std::ffi::CString;
+    use std::fs::{create_dir_all, remove_dir, remove_file};
+    use std::io::ErrorKind;
+    use std::os::unix::ffi::OsStrExt;
+    use std::os::unix::fs::symlink;
+
+    // Test non-existent path returns an error.
+    let c_path = CString::new("./nothing_to_see_here").expect("CString::new failed");
+    let r = unsafe { libc::realpath(c_path.as_ptr(), std::ptr::null_mut()) };
+    assert!(r.is_null());
+    let e = std::io::Error::last_os_error();
+    assert_eq!(e.raw_os_error(), Some(libc::ENOENT));
+    assert_eq!(e.kind(), ErrorKind::NotFound);
+
+    // Test that a long path returns an error.
+    //
+    // Linux first checks if the path to exists and macos does not.
+    // Using an existing path ensures all platforms return `ENAMETOOLONG` given a long path.
+    //
+    // Rather than creating a bunch of directories, we create two directories containing symlinks.
+    // Sadly we can't avoid creating directories via a path like "./././././" or"/../../" as linux
+    // appears to collapse "." and ".." before checking path length.
+    let path = tmp();
+
+    // The directories we will put symlinks in.
+    let x = path.join("x/");
+    let y = path.join("y/");
+
+    // The symlinks in each directory pointing to each other.
+    let yx_sym = y.join("x");
+    let xy_sym = x.join("y");
+
+    // Cleanup before test.
+    remove_file(&yx_sym).ok();
+    remove_file(&xy_sym).ok();
+    remove_dir(&x).ok();
+    remove_dir(&y).ok();
```

Comment:
It is probably better to copy the `prepare` functions from the `fs.rs` test; then you should just do this test in a subdirectory (`prepare_dir`) and don't have to remember to clean up everything separately.